Add emitting/listening support for metrics over the gRPC protocol

eriwo-stripe opened the pull request:
#### Summary

This PR adds the ability to listen for and emit metrics over the gRPC protocol. Both dogstatsd (metrics, events, sc's) and SFF are supported over gRPC. In addition this PR adds an option to veneur-emit that allows gRPC metrics to be proxyed before being sent to the server. We are also adding the ability to send dogstatsd over plain TCP in addition to the currently supported UDP.


#### Detailed Changelog

* Updated Changelog.md to include changes from this PR (woah I'm mentioning the changelog IN THE CHANGELOG. So meta.)
* Updated veneur-emit README to document the new gRPC flags and clarify the different protocols that veneur-emit supports.
* Update veneur-emit/main.go:
  * Added datadog TCP writer to write dogstatsd metrics to plain TCP
  * Added datadog gRPC writer to write dogstatsd metrics to plain gRPC
  * Updated event and service check logic to reduce code duplication and support gRPC
  * Added the `-grpc` and `-proxy` flags for sending metrics via gRPC
* Added proto schemas for sending SSF spans and dogstatsd metric packets 
* Updated generate.go to generate the protobuf files from the proto schema
* Updated networking.go to start gRPC server on the currently running healthcheck server
* Added unit tests to networking_test.go to test both dogstatsd and ssf gRPC metrics
* Updated server.go processMetricPacket() so that we can pass nil in place of the packet pool for metrics packets received over gRPC

#### Motivation

To enable metrics to be emitted/received over the gRPC protocol.


#### Test plan

* Added unit tests for both the SSF and dogstatsd gRPC servers
* Manual Testing
  * Local (Started server and manually verified that all emitted packets were received by the server): 
    * Start server: `go run ./cmd/veneur/main.go -f docs/dev.yaml`
    * Emit gRPC SSF spans: `go run ./cmd/veneur-emit/main.go -hostport tcp://localhost:8181 -debug -grpc -ssf -count 1337 -name eriwo.test.metric`
    * Emit dogstatsd metrics: `go run ./cmd/veneur-emit/main.go -hostport tcp://localhost:8181 -debug -grpc -count 1338 -name eriwo.test.metric`
    * Emit dogstatsd events: `go run ./cmd/veneur-emit/main.go -hostport tcp://localhost:8181 -debug -grpc -mode "event" -e_text "eriwo test event" -e_title "A test occurred"`
    * Emit dogstatsd service checks: `go run ./cmd/veneur-emit/main.go -hostport tcp://localhost:8181 -debug -grpc -mode "sc" -sc_status 0 -sc_name eriwo.test`
  * Remote:
    * Successfully emitted metrics over the chain of "local machine -> local SSL proxy -> remote load balancer -> remote SSL terminating envoy -> remote Veneur host -> SingalFX sink"
    * `go run ./cmd/veneur-emit/main.go -hostport loadbalancer.hostname.com:80 -proxy unix://$HOME/.localproxy -ssf -grpc -count 12 -name eriwo.test.metric -debug`
    * Verified that the test metric "eriwo.test.metric" was visible in the connected SignalFX


#### Rollout/monitoring/revert plan

This change should be monitored in two phases: initial deployment and metrics emitting migration.

1. During the initial deployment the change should be monitored to make sure that starting the gRPC servers does not negatively affect the host/network or cause any other problems
2. When clients begin emitting metrics over gRPC then it should be verified that metrics are being received without issue and are visible in any connected sinks

Rolling back this change for Veneur servers is as simple as reverting this commit and re-deploying.

If there is actually an error on the veneur-emit client side then the best way is not to roll back, but to roll forward. Assuming the actual API definition is right, we will just release another version with a fix rather than going to the effort to roll back on the client side.

If a revert of the veneut-emit client is absolutely necessary for some reason, then the clients need to change their command to not use the "-grpc" or "-proxy" flags and then this commit can be reverted/re-deployed. Another suggestion is to have an intermediate script between the clients and the server that wraps the client command in "-grpc"/"-proxy" so that you can just revert this script without touching the client command.

r? @rma-stripe @annab-stripe 
cc @stripe/observability

rma-stripe commented on cmd/veneur-emit/main.go:
```diff
@@ -241,9 +294,9 @@ func Main(args []string) int {
+		if !flagStruct.ToSSF && !flagStruct.ToGrpc {
 			logrus.WithField("ssf", flagStruct.ToSSF).
-				Error("Can't use tracing in non-ssf operation: Use -ssf to emit trace spans.")
+				Error("Can't use tracing in non-ssf operation: Use -ssf or -grpc to emit trace spans.")
```
I'm not sure I understand this comment and the surrounding logic. Veneur doesn't have support for spans (tracing) that are not formatted using the SSF format. SSF is a hard requirement, and GRPC is not, no?

eriwo-stripe replied:
Correct. I made this change early on when I didn't fully understand how this was working. You are right that it doesn't make sense so I will revert it.

rma-stripe reviewed (changes requested):
> Reverting this change for clients that are emitting is a little more tricky as the clients would be using (at minimum) the "-grpc" flag. If this is the case then the clients need to change their command to not use the "-grpc" or "-proxy" flags and then this commit can be reverted/re-deployed. Another suggestion is to have an intermediate script between the clients and the server that wraps the client command in "-grpc"/"-proxy" so that you can just revert this script without touching the client command.

I think in cases where a Veneur client changes its API, it makes sense for us to "roll forward" uses of the client rather than to "roll back" the changes (assuming we got the API design "right" on the latest revision).

I added some nits to this PR but overall, I think this will do what we want it to do. Let's walk over this PR file-by-file if you have a moment?

eriwo-stripe commented on cmd/veneur-emit/README.md:
```diff
@@ -94,6 +98,59 @@ instance.
+
+#### gRPC protocol:
+
+We support the gRPC protocol through the `-grpc` flag. When this flag is provided, we treat the `-hostport` argument as a TCP address or URL for the underlying gRPC connection.
```
Will update the wording here. I think saying "used in conjunction with" is redundant though as hostport is the main argument so everything is used in conjunction with it.

eriwo-stripe commented on cmd/veneur-emit/main.go:
```diff
@@ -176,49 +182,96 @@ func Main(args []string) int {
+	logrus.WithField("addr", addr).
+		WithField("net", netAddr.Network()).
+		WithField("destination", flagStruct.HostPort).
+		WithField("grcp", flagStruct.ToGrpc).
```
Nice catch!

eriwo-stripe commented on cmd/veneur-emit/main.go:
```diff
@@ -176,49 +182,96 @@ func Main(args []string) int {
+		return 1
+	}
+
+	logrus.WithField("addr", addr).
```
I agree and I actually prefer that was as well. However when I try to put it that way I get a compile error. I'm not familiar enough with go to know if this is actually a language requirement or if its just my IDE being weird.

eriwo-stripe commented on cmd/veneur-emit/README.md:
```diff
@@ -42,20 +42,24 @@ Usage of veneur-emit:
   -parent_span_id int
         ID of the parent span.
+  -proxy string
+        Address or URL to be used for proxing gRPC requests. The request will be sent to the proxy with the "hostport" attached so the proxy can forward the request. This can only be used with the "-grpc" flag.
```
Nice catch. Fixing.

eriwo-stripe commented on CHANGELOG.md:
```diff
@@ -4,7 +4,10 @@
 * Migrated from dep to Go modules. Clients must now use the updated import path `github.com/stripe/veneur/v14`. Thanks, [andybons](https://github.com/andybons)!
 
 ## Added
+* The ability to emit statsd metrics from veneur-emit over plain TCP in addition to the current plain UDP. Thanks, [shrivu-stripe](https://github.com/shrivu-stripe)!
```
You're right. I avoided bumping the version because I am not entirely sure what it entails. Now that all imports in go are "veneur/v14" does that mean we need to update every import to "v15"? Which files actually control the export artifacts version (is it just the dockerfile)?

I'll give it a shot now though and see how it goes.

eriwo-stripe replied:
After looking at [this article](https://golang.org/doc/modules/major-version) it seems like bumping the major version does require explicitly updating import tags in all files. I guess the easiest way is to not bump the major version anymore unless there actually are backwards incompatible changes. Therefore I'm going to add a new version of "v14.1.0"

eriwo-stripe commented:
Updated per Randall's comments. Bumped the version to v14.1.0, fixed comments/wording, and removed the unnecessary gRPC conditional.

r? @rma-stripe @annab-stripe

eriwo-stripe commented:
r? @rma-stripe @annab-stripe

annab-stripe reviewed (dismissed):
Great job!

rma-stripe commented on CHANGELOG.md:
```diff
@@ -1,11 +1,16 @@
 ## Updated
 * Migrated from dep to Go modules. Clients must now use the updated import path `github.com/stripe/veneur/v14`. Thanks, [andybons](https://github.com/andybons)!
 
-## Added
```
Note: this is red because this changelog entry was falsely added to a previous release.

eriwo-stripe replied:
![Shame](https://tenor.com/xrdS.gif)
